internal/engine/experiment/dnsscan/dnsscan.go

```diff
@@ -0,0 +1,151 @@
+// Package dnsscan is the experimental dnsscan experiment.
+package dnsscan
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"time"
+
+	"github.com/miekg/dns"
+	"github.com/ooni/probe-cli/v3/internal/measurexlite"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+)
+
+const (
+	testName    = "dnsscan"
+	testVersion = "0.1.0"
+)
+
+// TestKeys contains the experiment results.
+type TestKeys struct {
+	Queries        []*model.ArchivalDNSLookupResult `json:"queries"`
+	TargetResolver string                           `json:"target_resolver"`
+}
+
+// Config contains the experiment configuration.
+type Config struct {
+	// Address of the DNS resolver to be used for testing
+	Resolver string `ooni:"address of the DNS resolver to use"`
+}
+
+func (c Config) resolver() string {
+	if c.Resolver != "" {
+		return c.Resolver
+	}
+	return "udp://8.8.8.8:53"
+}
+
+// Measurer performs the measurement.
+type Measurer struct {
+	config Config
+}
+
+// ExperimentName implements ExperimentMeasurer.ExperiExperimentName.
+func (m *Measurer) ExperimentName() string {
+	return testName
+}
+
+// ExperimentVersion implements ExperimentMeasurer.ExperimentVersion.
+func (m *Measurer) ExperimentVersion() string {
+	return testVersion
+}
+
+var (
+	// errNoInputProvided indicates you didn't provide any input
+	errNoInputProvided = errors.New("not input provided")
+
+	// errInputFormatInvalid means that the input format is not valid
+	errInputFormatInvalid = errors.New("input is not a domain or URL")
+
+	// errNoInputProvided indicates that the chosen DNS resolver is not currently supported
+	errResolverNotSupported = errors.New("resolver not supported")
+
+	// errInvalidResolver indicates that the chosen DNS resolver is invalid
+	errInvalidResolver = errors.New("resolver address is invalid")
+
+	// errMissingPort indicates that there is no port.
+	errMissingPort = errors.New("the URL must include a port")
+)
+
+// Run implements ExperimentMeasurer.Run.
+func (m *Measurer) Run(
+	ctx context.Context,
+	sess model.ExperimentSession,
+	measurement *model.Measurement,
+	callbacks model.ExperimentCallbacks,
+) error {
+	if measurement.Input == "" {
+		return errNoInputProvided
+	}
+
+	// XXX do we also care to support URLs as inputs?
+	domain := string(measurement.Input)
+
+	tk := new(TestKeys)
+	tk.TargetResolver = m.config.resolver()
+	measurement.TestKeys = tk
+
+	resolver := m.config.resolver()
+	parsedResolver, err := url.Parse(string(resolver))
+	if err != nil {
+		return fmt.Errorf("%w: %s", errInvalidResolver, err.Error())
+	}
+	if parsedResolver.Scheme != "udp" {
+		return errResolverNotSupported
+	}
+	if parsedResolver.Port() == "" {
+		return errMissingPort
+	}
+	resolverIP := net.ParseIP(parsedResolver.Hostname())
+	if resolverIP == nil {
+		return fmt.Errorf("%w: invalid IP address", errInvalidResolver)
+	}
+	resolverAddress := fmt.Sprintf("%s:%s", resolverIP, parsedResolver.Port())
```

Comment:
```suggestion
	resolverAddress := parsedResolver.Host
```

Reply:
or, `net.JoinHostPort(resolverIP, parsedResolver.Port())` but you already have this value, so it's better to just use it from the URL rather than reconstructing it.